### PR TITLE
Adding Topic Creation And Subscription And Validating FCM Token

### PR DESCRIPTION
**Pull request by Stevemoretz**
Essentially I used : https://github.com/brozot/Laravel-FCM/pull/114
And brought it to your repo and made it work removed the errors and all,it works right now without any errors.

However I HAVE NOT TESTED the functionality completely yet.
I have however checked all the methods and it works without any errors.

### CODE REVIEW

File: src/Sender/FCMTopic.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        return false;
+    }
+
+    /**
+     * remove subscription from a topic.
+     *
+     *
+     * @param       $topic_id
+     * @param       $recipients_tokens
+     * @return null|string notification_key
```
Comment (williamdes) [inline src/Sender/FCMTopic.php]:
```suggestion
     * @return bool
```

File: src/Sender/FCMTopic.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+//        return $this->getNotificationToken($response);
+    }
+
+    /**
+     * add subscription to a topic.
+     *
+     * @param       $topic_id
+     * @param       $recipients_tokens
+     * @return null|string notification_key
```
Comment (williamdes) [inline src/Sender/FCMTopic.php]:
```suggestion
     * @return bool
```

File: src/Sender/FCMTopic.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    private $add_subscription_url = 'https://iid.googleapis.com/iid/v1:batchAdd';
+    private $remove_subscription_url = 'https://iid.googleapis.com/iid/v1:batchRemove';
+
+    /**
+     * Create a topic.
+     *
+     * @param       $notificationKeyName
+     * @param array $registrationIds
+     *
+     * @return null|string notification_key
```
Comment (williamdes) [inline src/Sender/FCMTopic.php]:
```suggestion
     * @return bool
```

**Comment by Stevemoretz** [comment]
Hi I'm new to this whole pull request thing, should I do anything now? Or it's okay?

I mean like confirming the changes you made?

**Comment by williamdes** [comment]
> Hi I'm new to this whole pull request thing, should I do anything now? Or it's okay?
> 
> I mean like confirming the changes you made?

Well, this seems okay
I will do the changes for you and add tests :)

If you want you can push more commits to add the changes I asked and tests before I do it

**Comment by Stevemoretz** [comment]
Well how do I add it?
There is a button under each one of your changes called 'Resolve Conversation'

The name make no sense, but if I press it, it will apply your changes?

I'm sorry again I'm new to this

**Comment by williamdes** [comment]
> Well how do I add it?

Do you have GIT setup for this repository ?
If yes, you need to clone your fork and checkout main branch, then you can push commits to here.

> The name make no sense, but if I press it, it will apply your changes?

It makes sense, this is indeed a discussion and you can click resolve if you think it is the case. In your case leave them as they are because I will fix them by pushing a commit to your pull-request.

**Comment by Stevemoretz** [comment]
I do have git setup for my repo.
`you need to clone your fork and checkout main branch`
but I don't know how to do this one,I usually do just push and pull and easy stuff I'm not familiar with these concepts that you are mentioning :(

**Comment by williamdes** [comment]
> I do have git setup for my repo.
> `you need to clone your fork and checkout main branch`
> but I don't know how to do this one,I usually do just push and pull and easy stuff I'm not familiar with these concepts that you are mentioning :(

Oh, okay

Then I will detail the steps to follow:
- `git clone git@github.com:Stevemoretz/Laravel-FCM.git Stevemoretz_Laravel-FCM` (you can use `https://github.com/Stevemoretz/Laravel-FCM.git` instead if your SSH key is not setup)
- `cd Stevemoretz_Laravel-FCM`
- `git checkout main` (checkouts the branch used for this PR, can be skipped because main is the default branch)
- Do you changes with any editor
- `git add -A` (adds all changes done)
- `git commit` (commits the changes)
- `git push` (Pushes the changes to this pull-request)`
- `git log --graph` (View the commit list)`

**Comment by Stevemoretz** [comment]
Thank you so much for the details.I understand the most parts of it so why would I do this I don't get it?
I already have it why would I clone it once again?
I already pushed it from my local repo.

I know how to fix my own repo I just don't get it how to bring the changed stuff here?
If I do another pull request then what happens to this one?Or that comes here too?

Wouldn't that be a problem if I do another pull request?

That's my problem I know how to push to my own repo but how to get it here I don't know?

**Comment by williamdes** [comment]
Then just push from your local repo and it will end up here :)

Local repositories are linked to remote ones by `git remote`

So, if you want to do the modification the do them, commit and push 
Nothing more ;)
Else just tell me and I will do them for you

PS: no need to close, open new pull-requests just push commits

**Comment by Stevemoretz** [comment]
Oh that easy just push :)
Sorry dude I didn't know that before so I'll apply your changes to my own repo and I'll let you know here when it's done.

**Comment by Stevemoretz** [comment]
Hi sorry for the delay I did your changes and pushed.

**Comment by Stevemoretz** [comment]
I also added some stuff for token validation and the method is tested and it works.

(It's only one method only).

File: src/Request/TopicRequest.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        parent::__construct();
+
+        if (!is_array($recipients_tokens)){
+            $recipients_tokens = [$recipients_tokens];
+        }
+
+        $this->topic_id = $topic_id;
+        $this->recipients_tokens = $recipients_tokens;
+        $this->operation = $operation;
```
Comment (williamdes) [inline src/Request/TopicRequest.php]: All this part is not in sync:
- Unused variables
- Missing variables
- phpdoc not done

File: README.md [hunk 5]
```diff
@@ -408,6 +433,13 @@ $topics->topic('TopicA')
        });
 ```
 
+## Validating
+
+### Validate FCM Token
+
+```
```
Comment (williamdes) [inline README.md]:
```suggestion
```php
```

**Comment by williamdes** [comment]
Hi @Stevemoretz 
Thank you for the changes, I added one more important comment.
All the other changes are okay

**Comment by Stevemoretz** [comment]
Hi @williamdes ,
Yeah sorry I didn't rewrite that code,I used : 

`https://github.com/brozot/Laravel-FCM/pull/114`

Anyways all fixed I guess.

**Comment by Stevemoretz** [comment]
Anyways what's the deal with these phpdocs at all?
I can't get it working,we use a facade so the php docs doesn't help the auto completion at all.
Either some helper php documentation should come in to fix this,or the whole thing should be changed to using directly classes and not a facade,otherwise no autocompletion which is wired.

**Comment by Stevemoretz** [comment]
For an example for this fix check FCMValidator class I added a php docs so my method : 
`FCMValidator::validateToken($validated['token']);`
Now simply gets an ide autocompletion because of the phpdoc on the facade.
```php
/**
 * @method static bool validateToken( $token )
 */
class FCMValidator extends Facade{
    protected static function getFacadeAccessor()
    {
        return 'fcm.validator';
    }
}

```

File: src/Request/TopicRequest.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace LaravelFCM\Request;
+
+/**
+ * Class GroupRequest.
```
Comment (williamdes) [inline src/Request/TopicRequest.php]:
```suggestion
 * Class TopicRequest.
```

File: src/Facades/FCMValidator.php [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace LaravelFCM\Facades;
+
+use Illuminate\Support\Facades\Facade;
+
+/**
+ * @method static bool validateToken( $token )
```
Comment (williamdes) [inline src/Facades/FCMValidator.php]: is this method needed ?

File: src/Request/TopicRequest.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+
+    private $topic_id;
+    /**
+     * @var array
+     */
+    private $recipients_tokens;
+
+    /**
+     * GroupRequest constructor.
```
Comment (williamdes) [inline src/Request/TopicRequest.php]:
```suggestion
     * TopicRequest constructor.
```

File: src/Request/TopicRequest.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param $operation
+     * @param $topic_id
```
Comment (williamdes) [inline src/Request/TopicRequest.php]:
```suggestion
     * @param string $operation
     * @param string $topic_id
```

File: src/Request/TopicRequest.php [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+class TopicRequest extends BaseRequest
+{
+    /**
+     * @internal
+     *
+     * @var string
+     */
+    protected $operation;
+
+
```
Comment (williamdes) [inline src/Request/TopicRequest.php]:
```suggestion

    /**
     * @internal
     *
     * @var string
     */
```

**Review by williamdes — APPROVED** [review]
I feel bad for requesting so much changes, I approve your work
:100:

**Comment by williamdes** [comment]
> For an example for this fix check FCMValidator class I added a php docs so my method :
> `FCMValidator::validateToken($validated['token']);`
> Now simply gets an ide autocompletion because of the phpdoc on the facade.
> 
> ```php
> /**
>  * @method static bool validateToken( $token )
>  */
> class FCMValidator extends Facade{
>     protected static function getFacadeAccessor()
>     {
>         return 'fcm.validator';
>     }
> }
> ```

Oh okay, maybe `@method static bool validateToken(string $token)` is better

**Comment by Stevemoretz** [comment]
Sure there you go.

Now I think the next step would be to fix other php docs of the facades.

**Comment by williamdes** [comment]
> Sure there you go.
> 
> Now I think the next step would be to fix other php docs of the facades.

Maybe phpstan is be the tool that we should use to be sure everything is fixed

**Comment by Stevemoretz** [comment]
> I feel bad for requesting so much changes, I approve your work
> 💯

No problem man Done all your other requests.

**Comment by Stevemoretz** [comment]
In a crazy cool way 
`barryvdh/laravel-ide-helper`
Actually writes the Facade php docs I am extremely fascinated.

**Comment by Stevemoretz** [comment]
Have never used phpstan though.

**Comment by williamdes** [comment]
> barryvdh/laravel-ide-helper

Yup, I used IDE helper and it works fine !
This is surely the response to https://github.com/code-lts/Laravel-FCM/pull/5#issuecomment-702139616

> Have never used phpstan though. 

Give it a try you will love it !

**Comment by williamdes** [comment]
I did lint fixes and added the author (@asachanfbd) back onto the commit and set ourselves to co-editors
I am planing to add tests before merging

**Comment by Stevemoretz** [comment]
> > barryvdh/laravel-ide-helper
> 
> Yup, I used IDE helper and it works fine !
> This is surely the response to [#5 (comment)](https://github.com/code-lts/Laravel-FCM/pull/5#issuecomment-702139616)
> 
> > Have never used phpstan though.
> 
> Give it a try you will love it !

Sure it is it magically created those php docs for me.

And Sure I'm right now installing phpstan thank you for introducing it!

**Comment by williamdes** [comment]
Could you have a try to see if the façades are okay on the commit f04fdc826456150a2a11303c94b7163b47c7b070 ?

You can `git fetch --all -p` and `git reset --hard f04fdc826456150a2a11303c94b7163b47c7b070` (if you have no unpushed changes)

**Comment by Stevemoretz** [comment]
> Could you have a try to see if the façades are okay on the commit [f04fdc8](https://github.com/code-lts/Laravel-FCM/commit/f04fdc826456150a2a11303c94b7163b47c7b070) ?
> 
> You can `git fetch --all -p` and `git reset --hard f04fdc826456150a2a11303c94b7163b47c7b070` (if you have no unpushed changes)

I tried it right now.Not really there isn't any phpdoc on FCMGroup for instance.

**Comment by williamdes** [comment]
> > Could you have a try to see if the façades are okay on the commit [f04fdc8](https://github.com/code-lts/Laravel-FCM/commit/f04fdc826456150a2a11303c94b7163b47c7b070) ?
> > You can `git fetch --all -p` and `git reset --hard f04fdc826456150a2a11303c94b7163b47c7b070` (if you have no unpushed changes)
> 
> I tried it right now.Not really there isn't any phpdoc on FCMGroup for instance.

Could you paste here the generated IDE helper part for FCM ?

**Comment by Stevemoretz** [comment]
Sure I guess it would be this part?

```php
    namespace LaravelFCM\Facades { 
            /**
     * 
     *
     */ 
        class FCM {
                    /**
         * send a downstream message to.
         * 
         * - a unique device with is registration Token
         * - or to multiples devices with an array of registrationIds
         *
         * @param string|array $to
         * @param \LaravelFCM\Sender\Options|null $options
         * @param \LaravelFCM\Sender\PayloadNotification|null $notification
         * @param \LaravelFCM\Sender\PayloadData|null $data
         * @return \LaravelFCM\Sender\DownstreamResponse|null 
         * @static 
         */ 
        public static function sendTo($to, $options = null, $notification = null, $data = null)
        {
                        /** @var \LaravelFCM\Sender\FCMSender $instance */
                        return $instance->sendTo($to, $options, $notification, $data);
        }
                    /**
         * Send a message to a group of devices identified with them notification key.
         *
         * @param string $notificationKey
         * @param \LaravelFCM\Sender\Options|null $options
         * @param \LaravelFCM\Sender\PayloadNotification|null $notification
         * @param \LaravelFCM\Sender\PayloadData|null $data
         * @return \LaravelFCM\Sender\GroupResponse 
         * @static 
         */ 
        public static function sendToGroup($notificationKey, $options = null, $notification = null, $data = null)
        {
                        /** @var \LaravelFCM\Sender\FCMSender $instance */
                        return $instance->sendToGroup($notificationKey, $options, $notification, $data);
        }
                    /**
         * Send message devices registered at a or more topics.
         *
         * @param \LaravelFCM\Sender\Topics $topics
         * @param \LaravelFCM\Sender\Options|null $options
         * @param \LaravelFCM\Sender\PayloadNotification|null $notification
         * @param \LaravelFCM\Sender\PayloadData|null $data
         * @return \LaravelFCM\Sender\TopicResponse 
         * @static 
         */ 
        public static function sendToTopic($topics, $options = null, $notification = null, $data = null)
        {
                        /** @var \LaravelFCM\Sender\FCMSender $instance */
                        return $instance->sendToTopic($topics, $options, $notification, $data);
        }
         
    }
            /**
     * 
     *
     */ 
        class FCMGroup {
                    /**
         * Create a group.
         *
         * @param string $notificationKeyName
         * @param array $registrationIds
         * @return null|string notification_key
         * @static 
         */ 
        public static function createGroup($notificationKeyName, $registrationIds)
        {
                        /** @var \LaravelFCM\Sender\FCMGroup $instance */
                        return $instance->createGroup($notificationKeyName, $registrationIds);
        }
                    /**
         * add registrationId to a existing group.
         *
         * @param string $notificationKeyName
         * @param string $notificationKey
         * @param array $registrationIds registrationIds to add
         * @return null|string notification_key
         * @static 
         */ 
        public static function addToGroup($notificationKeyName, $notificationKey, $registrationIds)
        {
                        /** @var \LaravelFCM\Sender\FCMGroup $instance */
                        return $instance->addToGroup($notificationKeyName, $notificationKey, $registrationIds);
        }
                    /**
         * remove registrationId to a existing group.
         * 
         * >Note: if you remove all registrationIds the group is automatically deleted
         *
         * @param string $notificationKeyName
         * @param string $notificationKey
         * @param array $registeredIds registrationIds to remove
         * @return null|string notification_key
         * @static 
         */ 
        public static function removeFromGroup($notificationKeyName, $notificationKey, $registeredIds)
        {
                        /** @var \LaravelFCM\Sender\FCMGroup $instance */
                        return $instance->removeFromGroup($notificationKeyName, $notificationKey, $registeredIds);
        }
                    /**
         * 
         *
         * @param \Psr\Http\Message\ResponseInterface $response
         * @return bool 
         * @static 
         */ 
        public static function isValidResponse($response)
        {
                        /** @var \LaravelFCM\Sender\FCMGroup $instance */
                        return $instance->isValidResponse($response);
        }
         
    }
     
}
```

**Comment by williamdes** [comment]
Yup, nothing for this PR ?
Or you would need me to merge so you can use the `dev-main` version of the composer package ?

**Comment by Stevemoretz** [comment]
> Yup, nothing for this PR ?
> Or you would need me to merge so you can use the `dev-main` version of the composer package ?

Well I think these features are fine now, but in other sections are still problems (not related to these though) like https://github.com/code-lts/Laravel-FCM/issues/7
And I still haven't tested it all so I think it's better you wait a little I test it then merging it would be a nice idea.

`dev-main version of the composer package `
I really don't know what this is :(

**Comment by williamdes** [comment]
> I really don't know what this is :(

Here you are :)
https://packagist.org/packages/code-lts/laravel-fcm#dev-main

**Comment by williamdes** [comment]
I merged this pull-request so you can test it on `dev-main`
I will maybe consider adding tests afterards

Anyway, thank you for all :)

**Comment by williamdes** [comment]
I added some tests in 1f5ac558083231451df1dab59d721135acb90b24

**Comment by williamdes** [comment]
I just released [1.6.0](https://github.com/code-lts/Laravel-FCM/releases/tag/1.6.0) :tada:

**Pull request merged**